datafusion/core/src/physical_optimizer/pipeline_checker.rs

```diff
@@ -0,0 +1,1173 @@
+///           +--------------+              +--------------+   - |  Hash Join   |-------| Projection   |
+///                                                            - |              |       |              |
+///           +--------------+              +--------------+  /  +--------------+       +--------------+
+///           |              |  unbounded   |              | /
+///    Right  | Finite       |    false     | Hash         |/false
+///           | Data Source  |--------------| Repartition  |
+///           |              |              |              |
+///           +--------------+              +--------------+
+///
+///
+///
+///           +--------------+              +--------------+
+///           |              |  unbounded   |              |
+///    Left   | Finite       |    false     | Hash         |\false
+///           | Data source  |--------------| Repartition  | \   +--------------+       +--------------+
+///           |              |              |              |  \  |              | true  |              | true
+///           +--------------+              +--------------+   - |  Hash Join   |-------| Projection   |-----
+///                                                            - |              |       |              |
+///           +--------------+              +--------------+  /  +--------------+       +--------------+
+///           |              |  unbounded   |              | /
+///    Right  | Infinite     |    true      | Hash         |/true
+///           | Data Source  |--------------| Repartition  |
+///           |              |              |              |
+///           +--------------+              +--------------+
+///
+/// ```
+fn hash_join_swap_subrule(
+    input: &PipelineStatePropagator,
+) -> Option<Result<PipelineStatePropagator>> {
+    let plan = input.plan.clone();
+    let children = &input.children_unbounded;
+    if let Some(hash_join) = plan.as_any().downcast_ref::<HashJoinExec>() {
+        let (left_unbounded, right_unbounded) = (children[0], children[1]);
+        let new_plan = match (left_unbounded, right_unbounded) {
+            (true, false) => {
+                if matches!(
+                    *hash_join.join_type(),
+                    JoinType::Inner
+                        | JoinType::Left
+                        | JoinType::LeftSemi
+                        | JoinType::LeftAnti
+                ) {
+                    swap(hash_join)
+                } else {
+                    Ok(plan)
+                }
+            }
+            _ => Ok(plan),
+        };
+        match new_plan {
+            Ok(plan) => Some(Ok(PipelineStatePropagator {
+                plan,
+                unbounded: left_unbounded || right_unbounded,
+                children_unbounded: vec![left_unbounded, right_unbounded],
+            })),
+            Err(e) => Some(Err(e)),
+        }
```

Comment:
Here is another way to do the same thing, which may or may not be more pleasing to your eyes:


```suggestion
        let new_state = new_plan.map(|plan| {
            PipelineStatePropagator {
                plan,
                unbounded: left_unbounded || right_unbounded,
                children_unbounded: vec![left_unbounded, right_unbounded],
            }
        });
        Some(new_state)
```